### ISSUE TITLE
ChainShape unpredictable behavior

### ISSUE THREAD

**Issue by dummyco** [erincatto/box2d#553]
Hello!
I'm getting weird `ChainShape` jittering behavior on sharp edges with joints on 2.3.1:

* `ChainShape` ground:
![2019-08-06_21-56-19](https://user-images.githubusercontent.com/3989453/62568173-51970380-b895-11e9-892a-a3f58e3552d4.gif)

* Simple `EdgeShape`s ground:
![2019-08-06_21-56-49](https://user-images.githubusercontent.com/3989453/62568172-51970380-b895-11e9-83df-f25e4520ed61.gif)

Any ideas on that?

Box2D scaling 16px to 1m, ~1-2m bounding box sizes for bodies.
0.5 densities for wheels, 5 density for the main body.
No dumping/hz joint tuning
-9.8 gravity, 1/60s step
8/4 iterations

**Comment by MelvMay-Unity** [comment]
It might not fix your issue but one small hack you could try is turning off the vertex0/vertex3 when getting edges if the continued edge is too acute or >90-deg.

As a test, try replacing the following function to see if it helps:
```
void b2ChainShape::GetChildEdge(b2EdgeShape* edge, int32 index) const
{
	b2Assert(0 <= index && index < m_count - 1);
	edge->m_type = b2Shape::e_edge;
	edge->m_radius = m_radius;

	edge->m_vertex1 = m_vertices[index + 0];
	edge->m_vertex2 = m_vertices[index + 1];

	if (index > 0)
	{
		edge->m_vertex0 = m_vertices[index - 1];
		edge->m_hasVertex0 = true;
	}
	else
	{
		edge->m_vertex0 = m_prevVertex;
		edge->m_hasVertex0 = m_hasPrevVertex;
	}

	if (index < m_count - 2)
	{
		edge->m_vertex3 = m_vertices[index + 2];
		edge->m_hasVertex3 = true;
	}
	else
	{
		edge->m_vertex3 = m_nextVertex;
		edge->m_hasVertex3 = m_hasNextVertex;
	}

	const b2Vec2 edgeDirection = edge->m_vertex2 - edge->m_vertex1;
	if (edge->m_hasVertex0)
	{
		const b2Vec2 adjacentDirection = edge->m_vertex0 - edge->m_vertex1;
		edge->m_hasVertex0 = b2Dot(-edgeDirection, adjacentDirection) >= -b2_epsilon;
	}
	if (edge->m_hasVertex3)
	{
		const b2Vec2 adjacentDirection = edge->m_vertex3 - edge->m_vertex2;
		edge->m_hasVertex3 = b2Dot(edgeDirection, adjacentDirection) >= -b2_epsilon;
	}
}
```

**Comment by erincatto** [comment]
Fixed in #607
